handlers.go

```diff
@@ -0,0 +1,132 @@
+package main
+
+import (
+	"database/sql"
+	"log"
+	"math/rand"
+	"net/http"
+	"strings"
+)
+
+const DefaultLang = "en-US"
+const DefaultOS = "win"
+
+// BouncerHandler is the primary handler for this application
+type BouncerHandler struct {
+	db *DB
+}
+
+func randomMirror(mirrors []MirrorsResult) *MirrorsResult {
+	totalRatings := 0
+	for _, m := range mirrors {
+		totalRatings += m.Rating
+	}
+	for _, m := range mirrors {
+		// Intn(x) returns from [0,x) and we need [1,x], so adding 1
+		rand := rand.Intn(totalRatings) + 1
+		if rand <= m.Rating {
+			return &m
+		}
+		totalRatings -= m.Rating
+	}
+
+	// This shouldn't happen
+	if len(mirrors) == 0 {
+		return nil
+	}
+	return &mirrors[0]
+}
+
+// URL returns the final redirect URL given a lang, os and product
+// if the string is == "", no mirror or location was found
+func (b *BouncerHandler) URL(lang, os, product string) (string, error) {
+	product, err := b.db.AliasFor(product)
+	if err != nil {
+		return "", err
+	}
+
+	osID, err := b.db.OSID(os)
+	switch {
+	case err == sql.ErrNoRows:
+		return "", nil
+	case err != nil:
+		return "", err
+	}
+
+	productID, sslOnly, err := b.db.ProductForLanguage(product, lang)
+	switch {
+	case err == sql.ErrNoRows:
+		return "", nil
+	case err != nil:
+		return "", err
+	}
+
+	locationID, locationPath, err := b.db.Location(productID, osID)
+	switch {
+	case err == sql.ErrNoRows:
+		return "", nil
+	case err != nil:
+		return "", err
+	}
+
+	mirrors, err := b.db.Mirrors(sslOnly, lang, locationID, true)
```

Comment:
mirrors presumably doesn't change too much, it it worth caching this value and having a goroutine update it periodically? 

I'm -1 on this idea if we can aggressively cache with nginx in front. If we put the logic here though we can remove an architectural layer